server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java

```diff
@@ -324,10 +324,17 @@ synchronized void cleanCache() {
         if (currentKeysToClean.isEmpty() == false || currentFullClean.isEmpty() == false) {
             for (Iterator<Key> iterator = cache.keys().iterator(); iterator.hasNext();) {
                 Key key = iterator.next();
+                CleanupKey cleanupKey = new CleanupKey(key.entity, key.readerCacheKey);
                 if (currentFullClean.contains(key.entity.getCacheIdentity())) {
+                    if (registeredClosedListeners.containsKey(cleanupKey)) {
+                        registeredClosedListeners.remove(cleanupKey);
+                    }
```

Comment:
Here is what I am thinking.

 - `registeredClosedListeners` are only present to clear the cache if the shard is closed. 
 - When closed, `cleanupKey.entity.isOpen() == false` is true, indicating a `currentFullClean`.
 -  The only other thing that indicates a `currentFullClean` is when `readerCacheKey == null`. `readerCacheKey == null` only happens on `IndicesRequestCache#clear`

So, it make sense that the two things that indicate a `currentFullClean` (meaning, removing all cache entries with the same cacheIdentity) also removes the `onClose` handler.

I need somebody else to confirm my thinking here. 

With all this said, I think the following changes.

 - Only call `registeredClosedListeners.remove(cleanupKey);` in the `if (currentFullClean.contains(key.entity.getCacheIdentity())) {` case. Do NOT do this in the `if (currentKeysToClean.contains(cleanupKey)) {`
 - In the `currentFullClean.contains(key.entity.getCacheIdentity())` you only need to call `registeredClosedListeners.remove(cleanupKey)`. Don't bother checking `registeredClosedListeners.containsKey(cleanupKey)`, that isn't necessary.